### DISCUSSION TITLE
Publish xmake to luarocks/PYPI for build lua/python module

### DISCUSSION THREAD

**Discussion by Freed-Wu** [xmake-io/xmake#5691] [Ideas]
Perhaps we can publish xmake to luarocks?

**Comment by Freed-Wu** [comment]
`xmake-scm-1.rockspec`:

```lua
local git_ref = '6c973f55a3130fe62342e4ab06e992eac6ee3a80'
local modrev = 'scm'
local specrev = '1'

local repo_url = 'https://github.com/xmake-io/xmake'

rockspec_format = '3.0'
package = 'xmake'
version = modrev ..'-'.. specrev

description = {
  summary = 'A cross-platform build utility based on Lua',
  detailed = '',
  labels = { 'c', 'linux', 'c-plus-plus', 'package-manager', 'visual-studio', 'cmake', 'build', 'lua', 'objective-c', 'makefile', 'build-tool', 'wdk', 'xmake', 'tbox', 'cross-toolchains' } ,
  homepage = 'https://luarocks.org/modules/waruqi/xmake',
  license = 'GPL-3.0'
}

dependencies = { 'lua >= 5.1' }

test_dependencies = { }

source = {
  url = repo_url .. '/archive/' .. git_ref .. '.zip',
  dir = 'xmake-' .. '6c973f55a3130fe62342e4ab06e992eac6ee3a80',
}

if modrev == 'scm' or modrev == 'dev' then
  source = {
    url = repo_url:gsub('https', 'git')
  }
end

build = {
  type = 'command',
  build_command = "./configure --prefix=$(PREFIX) && make",
  install_command = "make install"
}
```

**Reply by waruqi** [reply]
why? you can use https://github.com/xmake-io/luarocks-build-xmake

**Reply by Freed-Wu** [reply]
I means we can `luarocks install xmake`, not use xmake to build lua project :smile:

**Reply by waruqi** [reply]
This is unnecessary. Why not install it directly using the official script? Linux distributions have also provided it.

**Reply by Freed-Wu** [reply]
1. The release of luarocks can be controled by you to provide latest version every time. The release of other Linux distribution need to wait their package manager to update.
2. Provide a new release method will not be bad at least. In some speical case perhaps it can be the only one method to install for some users.

**Reply by waruqi** [reply]
why do not use `xmake update`?

```console
-- lastest version
xmake update 

-- dev version
xmake update dev

-- specific version
xmake update 2.9.5
```

you can run `source ~/.xmake/profile` to switch to these new versions

**Reply by Freed-Wu** [reply]
We can publish xmake to luarocks, provide binary rocks for usual platforms, then add it to dependencies of luarocks-build-xmake. When the user build a lua module which use xmake as its build sytem, it can save their effort to download and install xmake. all things are out of box. they don't need to know anything about which build system this lua module use.

**Reply by Freed-Wu** [reply]
This behaviour is usual. such as cmake and ninja, they are published to https://pypi.org/project/cmake/ and https://pypi.org/project/ninja/

If we want to use xmake to build python module in the future, we still publish xmake to PYPI, too, like cmake and ninja.

**Reply by waruqi** [reply]
This seemed like a good idea, but I didn't have much time to figure out how to publish it.

**Reply by Freed-Wu** [reply]
> figure out how to publish it

Add a CI/CD to do it automatically

**Reply by Freed-Wu** [reply]
For #5125, we need publish it to PYPI, npm, luarocks. That means we need to write pyproject.toml, package.json, xmake-scm-1.rockspec.
What is your preference about these files' location? simply put them in the root of project workspace?

**Reply by waruqi** [reply]
Usually they are maintained in independent repositories or third-party package management repositories.

https://github.com/Homebrew/homebrew-core/blob/master/Formula/x/xmake.rb

https://salsa.debian.org/xmake-team/xmake

**Reply by Freed-Wu** [reply]
However,

https://github.com/xmake-io/xmake/blob/dev/scripts/archlinux/PKGBUILD
https://github.com/xmake-io/xmake/blob/dev/scripts/rpmbuild/xmake.spec

Are in this project.

**Reply by waruqi** [reply]
> However,
> 
> https://github.com/xmake-io/xmake/blob/dev/scripts/archlinux/PKGBUILD https://github.com/xmake-io/xmake/blob/dev/scripts/rpmbuild/xmake.spec
> 
> Are in this project.

They are no longer maintained, the archlinux PKGBUILD is in the official repository and it will be maintained independently by the official maintainers.

https://gitlab.archlinux.org/archlinux/packaging/packages/xmake

The same is true for rpm packages. These files here are out of date.

**Reply by Freed-Wu** [reply]
> These files here are out of date.

These out-dated files should be removed.

>  be maintained independently by the official maintainers.

The difference is PYPI/npm/luarocks/... doesn't have any official maintainers.
You means any one who create a repo and register the name in PYPI/npm/... to do this?
At least you should own the privilege of managing the projects of PYPI/npm/...

**Reply by Freed-Wu** [reply]
This is an example, package esbuild to npm:

https://github.com/evanw/esbuild/tree/main/npm

**Reply by waruqi** [reply]
I have invited you to xmake-io group. you can create projects in https://github.com/xmake-io, then help to maintain these packages. e.g. xmake-pypi, xmake-npm, ...
